interfaces/builtin/gpio.go

```diff
@@ -0,0 +1,150 @@
+	if !ok {
+		return fmt.Errorf("gpio slot must have a number attribute")
+	}
+
+	// Valid values of number
+	if _, ok := number.(int); !ok {
+		return fmt.Errorf("gpio slot number attribute must be an int")
+	}
+
+	// Slot is good
+	return nil
+}
+
+// SanitizePlug checks the plug definition is valid
+func (iface *GpioInterface) SanitizePlug(plug *interfaces.Plug) error {
+	// Make sure right interface type
+	if iface.Name() != plug.Interface {
+		panic(fmt.Sprintf("plug is not of interface %q", iface))
+	}
+
+	// Plug is good
+	return nil
+}
+
+// PermanentPlugSnippet returns security snippets for plug at install
+func (iface *GpioInterface) PermanentPlugSnippet(plug *interfaces.Plug, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityDBus, interfaces.SecurityAppArmor, interfaces.SecuritySecComp, interfaces.SecurityUDev, interfaces.SecurityMount:
+		return nil, nil
+	default:
+		return nil, interfaces.ErrUnknownSecurity
+	}
+}
+
+// ConnectedPlugSnippet returns security snippets for plug at connection
+func (iface *GpioInterface) ConnectedPlugSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityDBus, interfaces.SecuritySecComp, interfaces.SecurityUDev, interfaces.SecurityMount:
+		return nil, nil
+	case interfaces.SecurityAppArmor:
+		path := fmt.Sprint(gpioSysfsGpioBase, slot.Attrs["number"])
+		// Entries in /sys/class/gpio for single GPIO's are just symlinks
+		// to their correct device part in the sysfs tree. As AppArmor
+		// does not handle symlinks we need to dereference the GPIO
+		// path and add the correct absolute path to the AppArmor snippet.
+		dereferencedPath, err := evalSymlinks(path)
+		if err != nil {
+			return nil, err
+		}
+		return []byte(fmt.Sprintf("%s/* rwk,\n", dereferencedPath)), nil
+	default:
+		return nil, interfaces.ErrUnknownSecurity
+	}
+}
+
+// PermanentSlotSnippet - no slot snippets provided
+func (iface *GpioInterface) PermanentSlotSnippet(slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityDBus, interfaces.SecurityAppArmor, interfaces.SecuritySecComp, interfaces.SecurityUDev, interfaces.SecurityMount:
+		return nil, nil
+	default:
+		return nil, interfaces.ErrUnknownSecurity
+	}
+}
+
+// ConnectedSlotSnippet - no slot snippets provided
+func (iface *GpioInterface) ConnectedSlotSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	// We need to export the GPIO so that it becomes as entry in sysfs
+	// available and we can assign it to a connecting plug.
+	number := []byte(strconv.Itoa(slot.Attrs["number"].(int)))
+	fileExport, err := os.OpenFile(gpioSysfsExport, os.O_WRONLY, 0200)
+	defer fileExport.Close()
+	if err != nil {
+		return nil, err
+	}
+	fileExport.Write(number)
```

Comment:
I've added an error check and as a result, added an extra test if the export had already taken place i.e. if /sys/class/gpio/gpioNUM symlink already exists. The file write does actually return an error if already exported so need to handle that error to allow connection to complete.